File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function (weights, days) {
    function ispossible(mid, weights, days) {
        let daycount=1;
        let sum=0;
        for(let i=0;i<weights.length;i++){
            if (sum+weights[i]<=mid){
                sum = sum + weights[i]

            }
            else{
                daycount++;
                sum = weights[i]
                if (daycount>days){
                    return false
                }
            }
        }
        return true

    }

    let low = Math.max(...weights);
    let high = weights.reduce((acc, current) => acc + current, 0);
    let res = -1;
    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        if (ispossible(mid, weights, days)) {
            res = mid;
            high = mid - 1;
        }
        else {
            low = mid + 1;
        }
    }
    return res;

};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool check(int mid, vector<int>& weights, int days, int n){
        int currentweight=0;
        int calculateddays =1;
        for(int i=0;i<n;i++){
            if(weights[i]>mid) return false;
            if(currentweight+ weights[i]<= mid){
                currentweight+=weights[i];
            }
            else{
                 currentweight=weights[i];
                 calculateddays++;

            }
        }
        if(calculateddays<=days) return true;
        return false;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int low = weights[0];
        int high =0;
        int n = weights.size();
        int ans =-1;

        for(int i=0;i<n;i++){
            high+= weights[i];
            low = max(low, weights[i]);
        }

        while(low<=high){
            int mid = (low)+ (high-low)/2;

            if(check(mid, weights, days,n)){
                ans = mid;
                high = mid-1;
            }
            else{
                low  = mid+1;
            }
        }

        return ans;
        
    }
};